File: dasc/ledger.py
import sqlite3
import json
import datetime
from .schemas import Decision, Intent

class BitemporalLedger:
    def __init__(self, db_path="dasc_ledger.db"):
        self.db_path = db_path
        self._init_db()

    def _init_db(self):
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                CREATE TABLE IF NOT EXISTS decisions (
                    id INTEGER PRIMARY KEY AUTOINCREMENT,
                    intent_id TEXT,
                    actor_agent TEXT,
                    status TEXT,
                    reason_codes TEXT,
                    intent_json TEXT,
                    timestamp TEXT,
                    previous_hash TEXT,
                    record_hash TEXT
                )
            """)

    def _get_last_hash(self):
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute("SELECT record_hash FROM decisions ORDER BY id DESC LIMIT 1")
            row = cursor.fetchone()
            return row[0] if row else "0" * 64

    def log_decision(self, intent: Intent, decision: Decision):
        from .sanitizer import sanitize_content
        from .utils import calculate_string_hash
        
        # Security: Scrub sensitive data from the intent JSON
        sanitized_intent_json = sanitize_content(intent.model_dump_json())
        prev_hash = self._get_last_hash()
        
        # Integrity: Calculate hash for this record including the previous hash
        record_content = f"{intent.intent_id}{decision.status}{sanitized_intent_json}{decision.timestamp}{prev_hash}"
        current_hash = calculate_string_hash(record_content)

        with sqlite3.connect(self.db_path) as conn:
            conn.execute("""
                INSERT INTO decisions (intent_id, actor_agent, status, reason_codes, intent_json, timestamp, previous_hash, record_hash)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """, (
                intent.intent_id,
                intent.actor_agent,
                decision.status,
                json.dumps(decision.reason_codes),
                sanitized_intent_json,
                decision.timestamp,
                prev_hash,
                current_hash
            ))
        print(f"[LEDGER] Recorded: {decision.status} for {intent.intent_id} (Hash: {current_hash[:8]})")

    def verify_integrity(self) -> bool:
        """
        Verifies the hash-chain integrity of the entire ledger.
        Returns False if any tampering is detected.
        """
        from .utils import calculate_string_hash
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute("SELECT * FROM decisions ORDER BY id ASC")
            columns = [column[0] for column in cursor.description]
            rows = [dict(zip(columns, row)) for row in cursor.fetchall()]
            
            expected_prev_hash = "0" * 64
            for row in rows:
                if row['previous_hash'] != expected_prev_hash:
                    return False
                
                # Re-calculate hash
                content = f"{row['intent_id']}{row['status']}{row['intent_json']}{row['timestamp']}{row['previous_hash']}"
                actual_hash = calculate_string_hash(content)
                if actual_hash != row['record_hash']:
                    return False
                
                expected_prev_hash = actual_hash
        return True

    def get_history(self):
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute("SELECT * FROM decisions ORDER BY timestamp DESC")
            columns = [column[0] for column in cursor.description]
            return [dict(zip(columns, row)) for row in cursor.fetchall()]

File: dasc-node/src/ledger.ts
import Database from "better-sqlite3";
import { Intent, Decision } from "./types.js";
import * as crypto from "crypto";

export class BitemporalLedger {
  private db: Database.Database;

  constructor(dbPath: string = "dasc_ledger_node.db") {
    this.db = new Database(dbPath);
    this.init();
  }

  private init() {
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS ledger (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        intent_id TEXT,
        actor_agent TEXT,
        status TEXT,
        reason_codes TEXT,
        intent_json TEXT,
        timestamp TEXT,
        previous_hash TEXT,
        record_hash TEXT
      )
    `);
  }

  private get_last_hash(): string {
    const row = this.db.prepare("SELECT record_hash FROM ledger ORDER BY id DESC LIMIT 1").get() as any;
    return row ? row.record_hash : "0".repeat(64);
  }

  private sanitize(content: string): string {
    // Basic PII scrubbing (SSN, Emails) for the audit trail
    return content
      .replace(/\b\d{3}-\d{2}-\d{4}\b/g, "[REDACTED_SSN]")
      .replace(/\b[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Z|a-z]{2,}\b/g, "[REDACTED_EMAIL]");
  }

  logDecision(intent: Intent, decision: Decision) {
    const prev_hash = this.get_last_hash();
    const intent_json = this.sanitize(JSON.stringify(intent));
    
    // Integrity: Calculate hash for this record including the previous hash
    const record_content = `${intent.intent_id}${decision.status}${intent_json}${decision.timestamp}${prev_hash}`;
    const record_hash = crypto.createHash("sha256").update(record_content).digest("hex");

    const stmt = this.db.prepare(`
      INSERT INTO ledger (intent_id, actor_agent, status, reason_codes, intent_json, timestamp, previous_hash, record_hash)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?)
    `);

    stmt.run(
      intent.intent_id,
      intent.actor_agent,
      decision.status,
      JSON.stringify(decision.reason_codes),
      intent_json,
      decision.timestamp,
      prev_hash,
      record_hash
    );
    
    console.log(`[Ledger] Recorded ${decision.status} for ${intent.intent_id} (Hash: ${record_hash.slice(0, 8)})`);
  }

  getHistory(limit: number = 50) {
    return this.db.prepare("SELECT * FROM ledger ORDER BY id DESC LIMIT ?").all(limit);
  }

  getHistoryAsOf(timestamp: string, limit: number = 50) {
    return this.db.prepare("SELECT * FROM ledger WHERE timestamp <= ? ORDER BY id DESC LIMIT ?").all(timestamp, limit);
  }

  verifyIntegrity(): boolean {
    const rows = this.db.prepare("SELECT * FROM ledger ORDER BY id ASC").all() as any[];
    let expected_prev_hash = "0".repeat(64);

    for (const row of rows) {
      if (row.previous_hash !== expected_prev_hash) return false;
      
      const content = `${row.intent_id}${row.status}${row.intent_json}${row.timestamp}${row.previous_hash}`;
      const actual_hash = crypto.createHash("sha256").update(content).digest("hex");
      
      if (actual_hash !== row.record_hash) return false;
      expected_prev_hash = actual_hash;
    }
    return true;
  }
}
